File: src/validate/naming_is_valid.rs
use crate::bundle_struct::Bundle;

pub async fn run(bundle: &Bundle) -> Result<(), String> {
    // Check method consistency
    let method = &bundle.method;
    let method_title = &bundle.ui.method.full_title;

    let expected_titles = match method.as_str() {
        "firehose" => vec![
            "Amazon Data Firehose",
            "AWS Firehose",
            "Kinesis Data Firehose",
        ],
        "s3" => vec!["Amazon S3", "AWS S3"],
        "kinesis" => vec!["Amazon Kinesis", "AWS Kinesis"],
        _ => vec![],
    };

    if !expected_titles.is_empty()
        && !expected_titles
            .iter()
            .any(|&title| method_title.contains(title))
    {
        return Err(format!(
            "ERROR: {}.{} docs.method.full_title '{method_title}' does not match method '{method}'",
            file!(),
            line!()
        ));
    }

    let source = &bundle.source;
    let source_title = &bundle.ui.source.full_title;

    if source == "waf" && !source_title.to_lowercase().contains("waf") {
        return Err(format!(
            "ERROR: {}.{} Source title should contain 'WAF' when source is 'waf'",
            file!(),
            line!()
        ));
    }

    // Check name consistency with source and method
    let name = &bundle.name;

    if !name.to_lowercase().contains(&source.to_lowercase())
        || !name.to_lowercase().contains(&method.to_lowercase())
    {
        return Err(format!(
            "ERROR: {}.{} Name '{name}' must include '{source}' and '{method}'",
            file!(),
            line!()
        ));
    }

    // Check version format (basic semantic versioning)
    let version = &bundle.metadata.version;
    if version.matches('.').count() != 2 {
        return Err(format!(
            "ERROR: {}.{} Version {version} should follow semantic versioning format (e.g., 1.0.0)",
            file!(),
            line!()
        ));
    }

    // Check maintainer email format
    let maintainer = &bundle.metadata.maintainer;
    if !maintainer.contains('@') || !maintainer.contains('.') {
        return Err(format!(
            "ERROR: {}.{} Maintainer should be a valid email address",
            file!(),
            line!()
        ));
    }

    // Check description is not empty
    if bundle.metadata.description.trim().is_empty() {
        return Err(format!(
            "ERROR: {}.{} Description cannot be empty",
            file!(),
            line!()
        ));
    }

    Ok(())
}

File: bundle-deployer/src/validation/naming_is_valid.ts
// Validation: Check that naming conventions are followed

import type { Bundle } from "../types/bundle.ts";

export function run(bundle: Bundle): void {
  // Check method consistency
  const method = bundle.method;
  const methodTitle = bundle.ui.method.full_title;

  const expectedTitles: Record<string, string[]> = {
    "firehose": [
      "Amazon Data Firehose",
      "AWS Firehose",
      "Kinesis Data Firehose",
    ],
    "s3": ["Amazon S3", "AWS S3"],
    "kinesis": ["Amazon Kinesis", "AWS Kinesis"],
  };

  const titles = expectedTitles[method] || [];
  
  if (titles.length > 0 && !titles.some(title => methodTitle.includes(title))) {
    throw new Error(
      `docs.method.full_title '${methodTitle}' does not match method '${method}'`
    );
  }

  // Check source consistency
  const source = bundle.source;
  const sourceTitle = bundle.ui.source.full_title;

  if (source === "waf" && !sourceTitle.toLowerCase().includes("waf")) {
    throw new Error("Source title should contain 'WAF' when source is 'waf'");
  }

  // Check name consistency with source and method
  const name = bundle.name;
  const nameLower = name.toLowerCase();

  if (!nameLower.includes(source.toLowerCase()) || 
      !nameLower.includes(method.toLowerCase())) {
    throw new Error(
      `Name '${name}' must include '${source}' and '${method}'`
    );
  }

  // Check version format (basic semantic versioning)
  const version = bundle.metadata.version;
  const dotCount = (version.match(/\./g) || []).length;
  
  if (dotCount !== 2) {
    throw new Error(
      `Version ${version} should follow semantic versioning format (e.g., 1.0.0)`
    );
  }

  // Check maintainer email format
  const maintainer = bundle.metadata.maintainer;
  if (!maintainer.includes('@') || !maintainer.includes('.')) {
    throw new Error("Maintainer should be a valid email address");
  }

  // Check description is not empty
  if (bundle.metadata.description.trim().length === 0) {
    throw new Error("Description cannot be empty");
  }
}